buidl/tx.py

```diff
@@ -88,6 +88,16 @@ def dump_cache(cls, filename):
             s = json.dumps(to_dump, sort_keys=True, indent=4)
             f.write(s)
 
+            
+class TxSender:
+    @classmethod
+    def sendrawtransaction(cls, signed_tx_hex_str, network="mainnet"):
+        """Broadcasts a signed transaction."""
+        url = "{}/tx".format(TxFetcher.get_url(network=network))
+        req = Request(url, data=signed_tx_hex_str.encode(), headers={"User-Agent": "Mozilla/5.0"})
+        response = urlopen(req)
+        return response.read().decode("utf-8") # returns the txid if successful
```

Comment:
Are you sure this code produced the results you cited? I tried submitting the TX hexes for the TXs you provided on testnet and signet and it failed for both (perhaps because the transactions in question are already in a block)?:

Signet `e940a16158bd2a75a13314c48175195d74e7a1782843cb04eef861621c2289e2`:
```
>>> from buidl.tx import TxSender
>>> TxSender.sendrawtransaction("01000000000101cfd5dc82017ecfd7da1987bc2061c4962b8f10cd2728bb163b52d5f782baa9240000000000ffffffff01ac840100000000001600143b608e0e95c8d8417feb5012394024ff24ad08d402483045022100c26b22cb7d664464353e72b29c1b78080df49b3db4f02655cbe2d99b9f1951cb02204278377d0d63c1cf3eb957ec449b0c11d27b09d3345253533decbdcacea4146d0121033bc7adaf5db7c1e528c6765f2ca28581f85c72a737a096457bc19d5b8206f18300000000", network="signet")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mflaxman/workspace/buidl-python/buidl/tx.py", line 98, in sendrawtransaction
    response = urlopen(req)
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 214, in urlopen
    return opener.open(url, data, timeout)
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 632, in http_response
    response = self.parent.error(
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 561, in error
    return self._call_chain(*args)
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request

```

Testnet `9f5f642021774524339e7845a0848f692e487f1752e578cfe79e34d616cd9287`:
```
>>> TxSender.sendrawtransaction("01000000000101d2e24e2866696273ce53e3513552f51cbf24ff5dfcf934d080390f7ca6c80c460000000000ffffffff015cc100000000000016001415c155d1eb6c370968ddac8256bb3cf4ad426f1502483045022100fd04b2081c9dc8404bd8ae1a5a82e5e84bde042a3e11b7a83fa4569a8c7ab6b20220328b7390ce6eb18184b465b285b040414854d7200b2adace3bc60417a2d88a680121023aa270709e17922cb5309ff76b89eb8cdb12daa11c1c62a6396fb8607ac5734e00000000", network="testnet")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/mflaxman/workspace/buidl-python/buidl/tx.py", line 98, in sendrawtransaction
    response = urlopen(req)
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 214, in urlopen
    return opener.open(url, data, timeout)
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 632, in http_response
    response = self.parent.error(
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 561, in error
    return self._call_chain(*args)
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/Users/mflaxman/.pyenv/versions/3.9.9/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/urllib/request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request
```

Assuming it's just buggy code, you can `grep` the code for `INCLUDE_NETWORK_TESTS` to see how to write tests that can be run locally but won't impact github.

Reply:
I believe the 400 would be expected, and that the guess is correct (failed because the transaction is already in a block).  The API call used in `sendrawtransaction()` broadcasts the signed transaction, so repeating after confirmation/spend should not succeed.  The exception falls through, to match the behavior of `TxFetcher.fetch()`, but this could be adjusted if desired.

Below was a test run using testnet that uses buidl to broadcast the signed transaction.  After confirmation, another attempt was made to broadcast the same transaction again (with 400 received).  Curl output (after confirmation) is also included.

```
>>> hd_priv = HDPrivateKey.from_mnemonic(mnemonic=test_mnemonic, network="testnet")
>>> utxo_txid = "1caec306cbafa06609b64bd2e5be11afef5c9733546caf20768b71922d268b25"
>>> utxo_index = 1
>>> tx_in = TxIn(bytes.fromhex(utxo_txid), utxo_index)
>>> fee = 500
>>> to_addr = "tb1qqy5y4dmvmfzjzwzr3jmjz45atm42eqrstmrxk7"
>>> amount = tx_in.value(network="testnet") - fee
>>> tx_out = TxOut.to_address(to_addr, amount)
>>> transaction = Tx(1, [tx_in], [tx_out], 0, network="testnet", segwit=True)
>>> utxo_key_path = "m/84h/1h/0h/0/0"
>>> utxo_key = hd_priv.traverse(utxo_key_path).private_key
>>> transaction.sign_input(0, utxo_key)
True
>>> TxFetcher.sendrawtransaction(transaction.serialize().hex(), network="testnet")
'f8c9fda4925ced18b227950bf953ff1ec60f1dc1c5d4ca6f3638a18ebee49e2e'
```
(after confirmation)
```
>>> TxFetcher.sendrawtransaction(transaction.serialize().hex(), network="testnet")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dev2/buidl-python/buidl/tx.py", line 96, in sendrawtransaction
    response = urlopen(req)
  File "/usr/lib/python3.9/urllib/request.py", line 214, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "/usr/lib/python3.9/urllib/request.py", line 632, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.9/urllib/request.py", line 561, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request
```
```
curl -X POST -w " Response Code: %{http_code}\\n" -sSLd "01000000000101258b262d92718b7620af6c5433975cefaf11bee5d24bb60966a0afcb06c3ae1c0100000000ffffffff015cc100000000000016001401284ab76cda452138438cb721569d5eeaac807002483045022100be180abb06b0856366e4564f824f9d96d7441b76877a934742b3bea6f06080db02202e3e937f096c2b6ddad48d3a811a04fb65dd177b4ad9be500e054e1a3e1e0ddf0121039b48c0c8fbe10f5ebd9a7175e17bfb0f4c484a23ca5d7359e4cb1821738f4fcd00000000" "https://blockstream.info/testnet/api/tx"
sendrawtransaction RPC error: {"code":-27,"message":"Transaction already in block chain"} Response Code: 400```
```


Below is a similar test on signet:
```
>>> hd_priv = HDPrivateKey.from_mnemonic(mnemonic=test2_mnemonic, network="signet")
>>> utxo_txid = "78f065fda37c93a2adb9189e8bf5d646b82f0042ef602d98a504f10948abae5c"
>>> utxo_index = 0
>>> tx_in = TxIn(bytes.fromhex(utxo_txid), utxo_index)
>>> fee = 500
>>> to_addr = "tb1qx55xuczv5644qpj99ughp6w8dq4lfl0j3hhfc7"
>>> amount = tx_in.value(network="signet") - fee
>>> tx_out = TxOut.to_address(to_addr, amount)
>>> transaction = Tx(1, [tx_in], [tx_out], 0, network="signet", segwit=True)
>>> utxo_key_path = "m/84h/1h/0h/0/0"
>>> utxo_key = hd_priv.traverse(utxo_key_path).private_key
>>> transaction.sign_input(0, utxo_key)
True
>>> TxFetcher.sendrawtransaction(transaction.serialize().hex(), network="signet")
'df5bf61946d5ea07b9c54fbe06b57b7054396ac0f0257a0949381e1ceb43e578'
```
(after confirmation)
```
>>> TxFetcher.sendrawtransaction(transaction.serialize().hex(), network="signet")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dev2/buidl-python/buidl/tx.py", line 96, in sendrawtransaction
    response = urlopen(req)
  File "/usr/lib/python3.9/urllib/request.py", line 214, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.9/urllib/request.py", line 523, in open
    response = meth(req, response)
  File "/usr/lib/python3.9/urllib/request.py", line 632, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.9/urllib/request.py", line 561, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 494, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.9/urllib/request.py", line 641, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request
```
```
curl -X POST -w " Response Code: %{http_code}\\n" -sSLd "010000000001015caeab4809f104a5982d60ef42002fb846d6f58b9e18b9ada2937ca3fd65f0780000000000ffffffff01ac8401000000000016001435286e604ca6ab5006452f1170e9c7682bf4fdf20247304402201ca8368d4ea48e269434b020778ac502bb07439ba41e11e0b5d9e3c5cba9fecd02206f31c3bc1aa05250504cc102deea2d900d7afd01e801b1c4e8379fd3fb6eb0bd012103a88b03985f2261a6a8d2263e6c62ddacdee3e78ebee70b7b637c78b881145b2500000000" "https://mempool.space/signet/api/tx"
sendrawtransaction RPC error: {"code":-27,"message":"Transaction already in block chain"} Response Code: 400
```

Reply:
Thanks @tdb3!

As long as the data gets passed back to the caller then a test of failing (resubmission of old hex) is good enough.

How about something like this?
```
from urllib.error import HTTPError
from urllib.request import Request, urlopen

url = "https://blockstream.info/testnet/api/tx"
tx_hex = "020000000001018f096e468b00d683d6e16015f71d1a5b4219b92526140447a59cedfe0155a1910000000000feffffff025898000000000000160014f1a37b9001e3fd8581129e6671ddeabd8d4b48d172c3050200000000160014311ac570af104db11cfb6e23fcaefb13df8876af02473044022032480522540adb4289857ca779777db79446e074edecae31e8b2cf4679910f3102207cc609a0e6d23b4f6bb0ac613aa828e5a4e36f208d47b4d1c23b098f55192d7d012103996c5401e4ed587ef3de9e4b35bdaf6dae34650c363c3c5a8fba2c2fd2eb8db8bab62300"
headers = {"User-Agent": "Mozilla/5.0"}

request = Request(url,  data=tx_hex.encode(), headers=headers)
try:
    print(urlopen(request).read().decode("utf-8"))
except HTTPError as err:
    err_details = f"{err.code} {err.msg}: {err.read().decode()}"
    raise ConnectionError(err_details)
```
returns:
```
ConnectionError: 400 Bad Request: sendrawtransaction RPC error: {"code":-27,"message":"Transaction already in block chain"}
```

Then you can write a `INCLUDE_NETWORK_TESTS` test for a duplicate submission that doesn't run on github (can't have any more dependencies) but is easy to run locally.

Sound good?

Reply:
Yeah, sounds good.  Appologies if I didn't fully understand, or am missing something.  I'm happy to tweak things to get it right.

Adjustment to `sendrawtransaction` was made (classmethod).

Two tests, `test_broadcast_duplicate_tx_spent_signet`, and `test_broadcast_duplicate_tx_spent_testnet` were added to test_tx.py (touches the blockstream.info and mempool.space API endpoints, for non-mainnet).  There were a few different approaches that could have been taken to create the tests.  Since `TxTest` seems to be focused primarily on testing `Tx` class functionality, a new class `TxFetcherTest` was created. Currently the new tests are the only tests in `TxFetcherTest`, but future unit tests could be added to `TxFetcherTest` (as needed).  `TxFetcherTest` inherits from OfflineTestCase (rather than TestCase) to to accommodate both offline and online tests (e.g., testing features similar to TxFetcher.fetch()'s caching capability). `test_broadcast_duplicate_tx_spent_signet/testnet` tests need network connectivity (i.e. with `INCLUDE_NETWORK_TESTS`), so skipUnless is used to prevent the tests from being run offline.

For `test_broadcast_duplicate_tx_spent_signet`, txid 78f065fda37c93a2adb9189e8bf5d646b82f0042ef602d98a504f10948abae5c was used.
For `test_broadcast_duplicate_tx_spent_testnet`, txid 1caec306cbafa06609b64bd2e5be11afef5c9733546caf20768b71922d268b25 was used.

The response from the API calls are:
400 Bad Request ... sendrawtransaction RPC error: {"code":-25,"message":"bad-txns-inputs-missingorspent"}
Although this error message covers multiple types of errors, the tests expect it when sending the duplicate spent transactions.

When sending duplicate unspent transactions, the API responses are:
400 Bad Request ... sendrawtransaction RPC error: {"code":-27,"message":"Transaction already in block chain"}
This response seems more ideal for tests (it's more specific to the case), but current understanding is that this would require those transactions to remain unspent indefinitely (preventing future use of the sats for other purposes without incurring test failures for buidl).  Tests using these UTXOs were avoided for this reason, but could be added easily if the benefit to buidl outweighs the marginal utility loss to testnet/signet (and smaller denominations could be used to minimize utility loss).


```
dev3@dev3:~/buidl-python$ python3 -m pytest -k broadcast -v
=========================================================== test session starts ============================================================
platform linux -- Python 3.9.2, pytest-6.2.5, py-1.11.0, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/dev3/buidl-python
collected 269 items / 267 deselected / 2 selected                                                                                          

buidl/test/test_tx.py::TxFetcherTest::test_broadcast_duplicate_tx_spent_signet PASSED                                                [ 50%]
buidl/test/test_tx.py::TxFetcherTest::test_broadcast_duplicate_tx_spent_testnet PASSED                                               [100%]

==================================================== 2 passed, 267 deselected in 1.51s =====================================================
```